### DISCUSSION TITLE
Sending POST to HTTPS results on "Could not create a secure connection"

### DISCUSSION THREAD

**Discussion by Hiranus** [cc-tweaked/CC-Tweaked#1991] [Q&A] [answered]
As is stated in topic. 
I want to create microservice on docker that I can send requests to. I can send requests via web browser just fine but when I try to do it with computercraft I get error
![image](https://github.com/user-attachments/assets/78444543-e406-4f31-b234-938122646e31)
(dont worry about WeatherForecast, that part is from template)
Request to the same address done via Swagger
![image](https://github.com/user-attachments/assets/e7de97d6-2c0d-4b51-beb7-8f947c522e89)

Here is my HTTP config. What should I do to fix this?
```
#Controls the HTTP API
[http]
	#Enable the "http" API on Computers. Disabling this also disables the "pastebin" and
	#"wget" programs, that many users rely on. It's recommended to leave this on and use
	#the "rules" config option to impose more fine-grained control.
	enabled = true
	#Enable use of http websockets. This requires the "http_enable" option to also be true.
	websocket_enabled = true
	#The number of http requests a computer can make at one time. Additional requests
	#will be queued, and sent when the running requests have finished. Set to 0 for
	#unlimited.
	#Range: > 0
	max_requests = 16
	#The number of websockets a computer can have open at one time.
	#Range: > 1
	max_websockets = 4

	#Limits bandwidth used by computers.
	[http.bandwidth]
		#The number of bytes which can be downloaded in a second. This is shared across all computers. (bytes/s).
		#Range: > 1
		global_download = 33554432
		#The number of bytes which can be uploaded in a second. This is shared across all computers. (bytes/s).
		#Range: > 1
		global_upload = 33554432

	#Tunnels HTTP and websocket requests through a proxy server. Only affects HTTP
	#rules with "use_proxy" set to true (off by default).
	#If authentication is required for the proxy, create a "computercraft-proxy.pw"
	#file in the same directory as "computercraft-server.toml", containing the
	#username and password separated by a colon, e.g. "myuser:mypassword". For
	#SOCKS4 proxies only the username is required.
	[http.proxy]
		#The type of proxy to use.
		#Allowed Values: HTTP, HTTPS, SOCKS4, SOCKS5
		type = "HTTP"
		#The hostname or IP address of the proxy server.
		host = ""
		#The port of the proxy server.
		#Range: 1 ~ 65536
		port = 8080

	#A list of rules which control behaviour of the "http" API for specific domains or
	#IPs. Each rule matches against a hostname and an optional port, and then sets several
	#properties for the request.  Rules are evaluated in order, meaning earlier rules override
	#later ones.
	#
	#Valid properties:
	# - "host" (required): The domain or IP address this rule matches. This may be a domain name
	#   ("pastebin.com"), wildcard ("*.pastebin.com") or CIDR notation ("127.0.0.0/8").
	# - "port" (optional): Only match requests for a specific port, such as 80 or 443.
	#
	# - "action" (optional): Whether to allow or deny this request.
	# - "max_download" (optional): The maximum size (in bytes) that a computer can download in this
	#   request.
	# - "max_upload" (optional): The maximum size (in bytes) that a computer can upload in a this request.
	# - "max_websocket_message" (optional): The maximum size (in bytes) that a computer can send or
	#   receive in one websocket packet.
	# - "use_proxy" (optional): Enable use of the HTTP/SOCKS proxy if it is configured.
	[[http.rules]]
		host = "localhost"
		action = "allow"

	[[http.rules]]
		#The maximum size (in bytes) that a computer can send or receive in one websocket packet.
		max_websocket_message = 131072
		host = "*"
		#The maximum size (in bytes) that a computer can upload in a single request. This
		#includes headers and POST text.
		max_upload = 4194304
		action = "allow"
		#Enable use of the HTTP/SOCKS proxy if it is configured.
		use_proxy = false
		#The maximum size (in bytes) that a computer can download in a single request.
		#Note that responses may receive more data than allowed, but this data will not
		#be returned to the client.
		max_download = 16777216
```

**Comment by piprett** [comment]
Is that microservice running with TLS/SSL? Have you tried using http as the protocol in the url instead of https? Is the certificate self-signed? What Minecraft version are you using?

**Reply by Hiranus** [reply]
It uses self-signed SSL.
MC version is 1.20.1
I will try to switch my project to http as a test, but I would prefer to have it as https if I can find a solution.

EDIT:
I tried http and it works

**Reply by SquidDev** [reply]
Ahh, if you're using a self-signed certificate, you'll need to add it to Java's certificate store. Though if you're just fetching from localhost, there's not much benefit for using https.

**Reply by Hiranus** [reply]
I switched to http and while get works without issue post complains 
![image](https://github.com/user-attachments/assets/d598ff9f-85cf-479b-9829-7697c2f63377)

Do you have any idea about this?

**Reply by SquidDev** [reply]
This is more specific to whatever code is running on your server, so afraid that's harder to answer without knowing more about the code there. 

I suspect you'll need to pass the content-type header (and possibly accept) too.

**Reply by Hiranus** [reply]
![image](https://github.com/user-attachments/assets/b0cc58f5-f659-402a-a193-980f5b989132)
I tried passing same header as is in swagger but I got nowhere
![image](https://github.com/user-attachments/assets/50d99d76-bcef-47b4-bd42-86b361a9fd98)
Even if I simplify endpoint to just accept string it still doesn't work.

as for what is running on docker - it's your ASP.NET Core Web API project, And it's super bare-bones as it's just Visual Studio template with one additional post endpoint
```C#
        [HttpPost]
        public void ItemsAsync([FromBody] Item items)
        {
        }
```

**Reply by fatboychummy** [reply]
`tostring(item)` -- Do you have an `__tostring` metamethod for converting your `item` object to a string? By default, doing `tostring` on any table-type object will just return `"table: ptr"`. I've not looked at the host code, but its possible it's expecting valid json data, and `table: xxxxxx` is invalid.

**Reply by Hiranus** [reply, marked as answer]
After some fiddling around with the code I got it to work. But I could swear that I tried it earlier and it didn't work... anyway, thanks for your help guys. SSL for HTTPS certificate was also a thing, so I will have to solve that at a later date, but for now I'll go with HTTP.
```lua
local header = {Accept = "*/*", ["Content-Type"] = "application/json"}
local rs = peripheral.wrap("bottom")
local rsItems = rs.listItems()
body = textutils.serializeJSON(rsItems[1])
http.post("http://localhost:32775/WeatherForecast",body,header).close()
```
